gobblin-utility/src/main/java/gobblin/util/PortUtils.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+
+import java.net.ServerSocket;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class PortUtils {
+  public static final int MINIMUM_PORT = 1025;
+  public static final int MAXIMUM_PORT = 65535;
+  private static final Pattern PORT_REGEX = Pattern.compile("\\$\\{PORT_(?:(\\?)|(\\d+)(\\+?))\\}");
+  private final PortLocator portLocator;
+
+  public PortUtils() {
+    this(new ServerSocketPortLocator());
+  }
+
+  @VisibleForTesting
+  PortUtils(PortLocator locator) {
+    this.portLocator = locator;
+  }
+
+  /**
+   * Replaces any port tokens in the specified string.
+   *
+   * NOTE: Tokens can be in the following forms:
+   *   1. ${PORT_123}
+   *   2. ${PORT_123+}
+   *   3. ${PORT_?}
```

Comment:
Can you explain what each pattern means, and what the code will do for each pattern.